### PR TITLE
Export anonymized Avro schema with schema backup tool

### PR DESCRIPTION
**Pull request by jjaakola-aiven**
# About this change - What it does

Added exporting of anonymized Avro schemas to schema backup tool.
Fixes the Avro part of issue #351.

# Why this way

The anonymization is done without the dependency to Avro library because main use case is to collect and export the existing schemas for running backwards compatibility test suite. Any incompatibility with schemas would cause the exporting to fail.

### CODE REVIEW

File: tests/integration/test_schema_backup_avro_export.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    # The export file has been created
+    assert os.path.exists(export_location)
+
+    expected_subject_found = False
+    with export_location.open("r") as fp:
+        exported_data = ujson.load(fp)
+        for msg in exported_data:
+            assert len(msg) == 2
+            key = msg[0]
+            if key.get("subject", None) == AVRO_SUBJECT_HASH:
```
Comment (jjaakola-aiven) [inline tests/integration/test_schema_backup_avro_export.py:85]: The test data from previous tests are in the schema registry topic. This check is for only asserting the expected Avro schema.

File: tests/integration/test_schema_backup_avro_export.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        json={"schema": f"{schema_string}", "schemaType": schemaType},
+    )
+    assert res.status == 200
+    assert "id" in res.json()
+
+
+async def test_export_anonymized_avro_schemas(
+    registry_async_client: Client, kafka_servers: KafkaServers, tmp_path: Path
+) -> None:
+    await insert_data(registry_async_client, "JSON", JSON_SUBJECT, JSON_SCHEMA)
```
Comment (jjaakola-aiven) [inline tests/integration/test_schema_backup_avro_export.py:68]: The JSON schema is not anonymized, it is skipped. Insert this data for asserting in the end of the test that it is not processed.

File: karapace/anonymize_schemas/anonymize_avro.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                union_types.append(anonymize_name(union_type))
+            else:
+                union_types.append(anonymize(union_type))
+        return union_types
+    return maybe_type_list_or_str
+
+
+def anonymize_schema(type_element: Union[str, Dict[str, str]], input_schema: Dict[str, Any]) -> Dict[str, Any]:
+    schema: Dict[str, Any] = {}
+    for key, value in input_schema.items():
```
Comment (hackaugusto) [inline karapace/anonymize_schemas/anonymize_avro.py:67]: This loop is not handling [logical types](https://avro.apache.org/docs/current/spec.html#Logical+Types).

File: karapace/anonymize_schemas/anonymize_avro.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                union_types.append(union_type)
+            elif isinstance(union_type, str):
+                union_types.append(anonymize_name(union_type))
+            else:
+                union_types.append(anonymize(union_type))
+        return union_types
+    return maybe_type_list_or_str
+
+
+def anonymize_schema(type_element: Union[str, Dict[str, str]], input_schema: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (hackaugusto) [inline karapace/anonymize_schemas/anonymize_avro.py:65]: Suggestion:

```suggestion
def anonymize_complex_type(type_element: Union[str, Dict[str, str]], input_schema: Dict[str, Any]) -> Dict[str, Any]:
```

File: karapace/anonymize_schemas/anonymize_avro.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    Returns anonymized name.
+    """
+    anonymized_elements = []
+    for element in name.split("."):
+        element = hashlib.sha1(element.encode("utf-8")).hexdigest()
+        # SHA-1 can start with number, just add 'a' as first character.
+        # This breaks the hash, but is still consistent as required.
+        as_list_element = list(element)
+        as_list_element[0] = "a"
+        anonymized_elements.append("".join(as_list_element))
```
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply karapace/anonymize_schemas/anonymize_avro.py:46]: The intent is also to break the hash.

File: karapace/anonymize_schemas/anonymize_avro.py (hunk 4 above)
Comment (jjaakola-aiven) [reply karapace/anonymize_schemas/anonymize_avro.py]: Done.

File: karapace/anonymize_schemas/anonymize_avro.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            else:
+                # Nested type
+                schema[key] = anonymize(value)
+
+    return schema
+
+
+def anonymize(input_schema: Union[str, Dict[str, Any], List]) -> Union[str, Dict[str, Any], List]:
+    if not input_schema:
+        return {}
```
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply karapace/anonymize_schemas/anonymize_avro.py]: Done.

File: karapace/anonymize_schemas/anonymize_avro.py (hunk 3 above)
Comment (jjaakola-aiven) [reply karapace/anonymize_schemas/anonymize_avro.py:74]: Done.

File: karapace/anonymize_schemas/anonymize_avro.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        # SHA-1 can start with number, just add 'a' as first character.
+        # This breaks the hash, but is still consistent as required.
+        as_list_element = list(element)
+        as_list_element[0] = "a"
+        anonymized_elements.append("".join(as_list_element))
+
+    return ".".join(anonymized_elements)
+
+
+def anonymize_type(maybe_type_list_or_str: Union[str, List[Any]]) -> Union[str, List[Any]]:
```
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply karapace/anonymize_schemas/anonymize_avro.py]: Done.

File: karapace/anonymize_schemas/anonymize_avro.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    "timestamp-micros",
+    "timestamp-millis",
+    "uuid",
+]
+ALL_TYPES = PRIMITIVE_TYPES + LOGICAL_TYPES + ["array", ENUM, "fixed", "map", "record"]
+
+ORDER_VALID_VALUES = ["ascending", "descending", "ignore"]
+
+
+def anonymize_name(name: str) -> str:
```
Comment (jjaakola-aiven) [inline karapace/anonymize_schemas/anonymize_avro.py:64]: This function assumes that names are always valid in the input data.
If there are invalid names the algorithm would need changes:

* Split by dot.
* For each element.
    * Hash the string.
    * Add the first character from the original string to start of the hash.
    * Collect every character from the original string not matching to the Avro name rule and add to the end of the string
* Concatenate elements by dot to single string.

After these steps the anonymized string contains also the invalid characters.

**Review by hackaugusto — APPROVED** [review]
LGTM

**Pull request merged**